posts/2021/01/pytorch-tensoriterator-internals-update.md

```diff
@@ -0,0 +1,264 @@
+<!--
+.. title: PyTorch TensorIterator Internals (TensorIteratorConfig Update)
+.. slug: pytorch-tensoriterator-internals-update
+.. date: 2021-01-27 00:40:00 UTC-06:00
+.. author: Kurt Mohler
+.. tags: 
+.. category: 
+.. link: 
+.. description: 
+.. type: text
+-->
+
+For contributors to the PyTorch codebase, one of the most commonly encountered
+C++ classes is
+[`TensorIterator`](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/TensorIterator.h).
+`TensorIterator` offers a standardized way to iterate over elements of
+a tensor, automatically parallelizing operations, while abstracting device and 
+data type details.
+
+In April 2020, Sameer wrote a blog article discussing [PyTorch TensorIterator Internals](../../2020/04/pytorch-tensoriterator-internals.html). Recently,
+however, the interface has changed significantly. This post describes how to
+use the current interface as of January 2021. Much of the information from the
+previous article is directly copied here, but with updated API calls and some
+extra details.
+
+All of the code examples below can be compiled and run in [this GitHub
+repo](https://github.com/kurtamohler/pytorch-TensorIterator-examples).
+
+# Basics of TensorIterator and TensorIteratorConfig
+
+In order to create a `TensorIterator`, a `TensorIteratorConfig` must be created
+first. `TensorIteratorConfig` specifies the input and output tensors that will
```

Comment:
A fun or a not so fun fact, `TensorIteratorConfig` is not copyable nor copy-assignable :(